sys/kern/vm_map.c

```diff
@@ -161,33 +165,53 @@ void vm_map_entry_destroy(vm_map_t *map, vm_map_entry_t *ent) {
   vm_map_entry_free(ent);
 }
 
+static inline vm_map_entry_t *vm_map_entry_copy(vm_map_entry_t *src) {
+  if (src->object)
+    vm_object_hold(src->object);
+  vm_map_entry_t *new = vm_map_entry_alloc(src->object, src->start, src->end,
+                                           src->prot, src->flags);
+  return new;
+}
+
+/* Split vm_map_entry into two not empty entries. (Smallest possible entry is
+ * entry with one page thus splitat must be page aligned.)
+ *
+ * Returns entry which is after base entry. */
+static vm_map_entry_t *vm_map_entry_split(vm_map_t *map, vm_map_entry_t *ent,
+                                          vaddr_t splitat) {
+  assert(mtx_owned(&map->mtx));
+  assert(page_aligned_p(splitat));
+  assert(ent->start < splitat && splitat < ent->end);
+
+  vm_map_entry_t *new_ent = vm_map_entry_copy(ent);
+
+  /* clip both entries */
```

Comment:
I wonder... shouldn't we have `vm_map_entry::offset` that helps to determine which part of vm object is actually visible through this map entry?

Let's say I have a map entry `A(start=va+0,size=16384,vmobj=VO)`. I split it into `B(start=va+0,size=4096,vmobj=VO)` and `C(start=va+8192,size=8192,vmobj=VO)`. How page fault routine will know for `va+8192` that it should look up a page in `VO` at offset 8192, instead of 0?

Reply:
Yes, we need that field. I'va added it and test to check if it works.